src/compile/signal.ts

```diff
@@ -1,56 +1,33 @@
+import {SignalRef} from 'vega';
 import {codegen, parse} from 'vega-expression';
-import {stringValue} from 'vega-util';
-import {globalWholeWordRegExp, keys} from '../util';
+
+export type Rename = (oldSignalName: string) => string;
 
 /**
- * A class that wraps an expression with a list signal names in the expression. If they are renamed during parsing multi-views,   we can rename then signals in the assemble phase.
+ * A class that behaves like a SignalRef but lazily generates the signal.
+ * The provided generator function should use `Model.getSignalName` to use the correct signal name.
  */
-export class SignalRefComponent {
-  constructor(public expr: string, public signalNames: string[]) {}
+export class SignalRefWrapper implements SignalRef {
+  constructor(private exprGenerator: () => string) {}
 
-  public static fromName(signalName: string) {
-    return new SignalRefComponent(signalName, [signalName]);
+  public get signal() {
+    return this.exprGenerator();
   }
-}
-
-const generate = codegen({globalvar: 'global'});
 
-export function evalOrMakeSignalRefComponent(
-  expr: string,
-  params: {[varname: string]: number | string | boolean | object | SignalRefComponent}
-) {
-  const varNames: string[] = keys(params);
-  const signalNames = [];
+  public toJSON() {
+    return {signal: this.signal};
+  }
 
-  for (const varName of varNames) {
-    const param = params[varName];
-    if (param instanceof SignalRefComponent) {
-      for (const name of param.signalNames) {
-        signalNames.push(name);
-      }
-      expr = expr.replace(globalWholeWordRegExp(varName), param.expr);
-    } else {
-      // primitive value
-      expr = expr.replace(globalWholeWordRegExp(varName), stringValue(param));
-    }
+  public static fromName(rename: Rename, signalName: string) {
+    return new SignalRefWrapper(() => rename(signalName));
   }
+}
 
-  if (signalNames.length > 0) {
-    return new SignalRefComponent(expr, signalNames);
-  } else {
-    try {
-      // Try to evaluate the expression and return the value if we succeed.
+const generate = codegen({globalvar: 'global'});
 
-      const ast = parse(expr);
-      const {code} = generate(ast);
-      const f = new Function('global', `return ${code};`); // tslint:disable-line:function-constructor
-      return f(params);
-    } catch (error) {
-      if (error.message.indexOf('Unrecognized function') === 0) {
-        // This expression contains a function that needs to be evaluated at runtime, so let's keep it as a SignalRefComponent.
-        return new SignalRefComponent(expr, []);
-      }
-      throw error;
-    }
-  }
+export function evalExpression(expr: string) {
+  const ast = parse(expr);
+  const {code} = generate(ast);
+  const f = new Function('global', `return ${code};`); // tslint:disable-line:function-constructor
```

Comment:
There are some expression that need to be run at runtime, and that will fail this `new Function` thing, so we better keep the try-catch here.

Reply:
Okay. I wonder whether we should just keep the expressions. How much do we gain by compiling them into values?

Reply:
Btw, I think you are wrong. The typings say that we will only ever have numbers as unknown parameters when we eval (for Signals we always generate the expression).